src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs

```diff
@@ -57,8 +66,27 @@ async Task<CompletableSynchronizedStorageSession> AdaptOrOpenNewSynchronizedStor
                    ?? await synchronizedStorage.OpenSession(contextBag).ConfigureAwait(false);
         }
 
+        static void LogHandlersInvocation(IIncomingLogicalMessageContext context, List<MessageHandler> handlersToInvoke)
+        {
+            var builder = new StringBuilder();
+            builder.AppendLine($"Processing message type: {context.Message.MessageType}");
+            builder.AppendLine($"Headers: {Environment.NewLine}{string.Join(", ", context.Headers.Select(h => $"{h.Key}:{h.Value}").ToArray())}");
+            builder.Append("Handlers to invoke: ");
+
+            foreach (var messageHandler in handlersToInvoke)
+            {
+                builder.Append($"{messageHandler.HandlerType.FullName} => ");
+            }
+
+            builder.Length -= 4;
+
+            logger.Debug(builder.ToString());
+        }
+
         readonly ISynchronizedStorageAdapter adapter;
         readonly ISynchronizedStorage synchronizedStorage;
         readonly MessageHandlerRegistry messageHandlerRegistry;
+        static readonly ILog logger = LogManager.GetLogger<LoadHandlersConnector>();
+        static readonly bool isDebugIsEnabled = logger.IsDebugEnabled;
```

Comment:
this can still be removed

Reply:
Disagree. See https://github.com/Particular/NServiceBus/pull/5230#discussion_r209986647
Perhaps @Particular/nservicebus-maintainers should agree among themselves first? 😉

Reply:
it seems like in the codebase both approaches are used. Not really a big fan of it tbh (especially as technically, it's easy to make classes cache an incorrect log level in case somebody configured the log level somewhere in between the endpoint configuration or later) but I'll retract my comment based on that.